storagenode/pieces/readwrite.go

```diff
@@ -32,28 +88,84 @@ func NewWriter(blob storage.BlobWriter) (*Writer, error) {
 // Write writes data to the blob and calculates the hash.
 func (w *Writer) Write(data []byte) (int, error) {
 	n, err := w.blob.Write(data)
-	w.size += int64(n)
+	w.pieceSize += int64(n)
 	_, _ = w.hash.Write(data[:n]) // guaranteed not to return an error
 	if err == io.EOF {
 		return n, err
 	}
 	return n, Error.Wrap(err)
 }
 
-// Size returns the amount of data written so far.
-func (w *Writer) Size() int64 { return w.size }
+// Size returns the amount of data written to the piece so far, not including the size of
+// the piece header.
+func (w *Writer) Size() int64 { return w.pieceSize }
 
 // Hash returns the hash of data written so far.
 func (w *Writer) Hash() []byte { return w.hash.Sum(nil) }
 
 // Commit commits piece to permanent storage.
-func (w *Writer) Commit(ctx context.Context) (err error) {
+func (w *Writer) Commit(ctx context.Context, pieceHeader *pb.PieceHeader) (err error) {
 	defer mon.Task()(&ctx)(&err)
 	if w.closed {
 		return Error.New("already closed")
 	}
+
+	// point of no return: after this we definitely either commit or cancel
 	w.closed = true
-	return Error.Wrap(w.blob.Commit(ctx))
+	defer func() {
+		if err != nil {
+			err = Error.Wrap(errs.Combine(err, w.blob.Cancel(ctx)))
+		} else {
+			err = Error.Wrap(w.blob.Commit(ctx))
+		}
+	}()
+
+	formatVer := w.blob.GetStorageFormatVersion()
+	pieceHeader.FormatVersion = int32(formatVer)
+	if formatVer == storage.FormatV0 {
+		return nil
+	}
+	headerBytes, err := proto.Marshal(pieceHeader)
+	if err != nil {
+		return err
+	}
+	if len(headerBytes) > (V1PieceHeaderReservedArea - v1PieceHeaderFramingSize) {
+		// This should never happen under normal circumstances, and it might deserve a panic(),
+		// but I'm not *entirely* sure this case can't be triggered by a malicious uplink. Are
+		// google.protobuf.Timestamp fields variable-width?
+		return Error.New("marshaled piece header too big!")
+	}
+	size, err := w.blob.Size()
+	if err != nil {
+		return err
+	}
+	if _, err := w.blob.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+
+	// We need to store some "framing" bytes first, because protobufs are not self-delimiting.
+	// In cases where the serialized pieceHeader is not exactly V1PieceHeaderReservedArea bytes
+	// (probably _all_ cases), without this marker, we wouldn't have any way to take the
+	// V1PieceHeaderReservedArea bytes from a piece blob and trim off the right number of zeroes
+	// at the end so that the protobuf unmarshals correctly.
+	var framingBytes [v1PieceHeaderFramingSize]byte
+	binary.BigEndian.PutUint16(framingBytes[:], uint16(len(headerBytes)))
```

Comment:
Probably overkill here, but a few comments. First, it appears the choice of uint16 is dependent on the `v1PieceHeaderFramingSize`, even though they seem unrelated.

Second, do we want to check thta `len(headerBytes)` fits within a uint16? I suppose that will get caught in the check above when we check whether the entire length fits within the reserved space, but I'm just thinking of ways a future change could cause problems, if that makes sense.

Reply:
Yep, the uint16 is dependent on v1PieceHeaderFramingSize. If this were c++, I'd certainly define v1PieceHeaderFramingSize in terms of the uint16 type, and use the `sizeof` operator. There doesn't appear to be any way to do that cleanly in Go. You have to call `PutUint16` by name. As far as I can tell, this is what everybody else does in Go land when they have int field sizes like this; they assume that people can recognize that the size of the type is related to the size of the values it stores. I think having the framingBytes slice created on the line immediately before ought to be a pretty good hint.

I really hope we don't ever have piece headers of more than 64kiB, but if we ever do, I'm sure that we'll need to reexamine a lot of things besides this framing field, so I'm ok with the amount of sanity checking I think.

Reply:
Fair enough.